src/MSPFormat.jl

```
@@ -0,0 +1,297 @@
+#  Copyright (c) 2017-23, Oscar Dowson and SDDP.jl contributors
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+module MSPFormat
+
+import JSON
+import JuMP
+import ..SDDP
+
+function _parse_lattice(filename::String)
+    data = JuMP.MOI.FileFormats.compressed_open(
+        filename,
+        "r",
+        JuMP.MOI.FileFormats.AutomaticCompression(),
+    ) do io
+        return JSON.parse(io)
+    end
+    graph = SDDP.Graph("root")
+    for key in keys(data)
+        SDDP.add_node(graph, key)
+    end
+    for (key, value) in data
+        for (child, probability) in value["successors"]
+            SDDP.add_edge(graph, key => child, probability)
+        end
+    end
+    # MSPFormat doesn't have explicit root -> stage 1 arcs. Assume uniform.
+    # Also, MSPFormat uses 0-indexed stages.
+    stage_zero = String[key for (key, value) in data if value["stage"] == 0]
+    for key in stage_zero
+        SDDP.add_edge(graph, "root" => key, 1 / length(stage_zero))
+    end
+    return graph, data
+end
+
+# If the state is missing, the default is `0.0`.
+_get_constant(terms::String, state::Dict) = get(state, terms, 0.0)
```

This should throw an error if the term isn't present